capsules/src/led_matrix.rs

```diff
@@ -0,0 +1,247 @@
+//!     ]
+//! );
+//!
+//! let led = static_init!(
+//!     capsules::led_matrix::LedMatrixDriver<
+//!         'static,
+//!         nrf52::gpio::GPIOPin<'static>,
+//!         capsules::virtual_alarm::VirtualMuxAlarm<'static, nrf52::rtc::Rtc<'static>>,
+//!     >,
+//!     capsules::led_matrix::LedMatrixDriver::new(cols, rows, buffer, led_alarm, kernel::hil::gpio::ActivationMode::ActiveLow, kernel::hil::gpio::ActivationMode::ActiveHigh, 60)
+//! );
+//!
+//! led_alarm.set_alarm_client(led);
+//!
+//! led.init();
+//! ```
+//!
+//! Syscall Interface
+//! -----------------
+//!
+//! - Stability: 2 - Stable
+//!
+//! ### Command
+//!
+//! All LED operations are synchronous, so this capsule only uses the `command`
+//! syscall.
+//!
+//! #### `command_num`
+//!
+//! - `0`: Return the number of LEDs on this platform.
+//!   - `data`: Unused.
+//!   - Return: Number of LEDs.
+//! - `1`: Turn the LED on.
+//!   - `data`: The index of the LED. Starts at 0.
+//!   - Return: `SUCCESS` if the LED index was valid, `EINVAL` otherwise.
+//! - `2`: Turn the LED off.
+//!   - `data`: The index of the LED. Starts at 0.
+//!   - Return: `SUCCESS` if the LED index was valid, `EINVAL` otherwise.
+//! - `3`: Toggle the on/off state of the LED.
+//!   - `data`: The index of the LED. Starts at 0.
+//!   - Return: `SUCCESS` if the LED index was valid, `EINVAL` otherwise.
+
+use kernel::hil::gpio;
+use kernel::{AppId, Driver, ReturnCode};
+
+use core::cell::Cell;
+use kernel::common::cells::TakeCell;
+
+use kernel::hil::gpio::ActivationMode;
+use kernel::hil::time::{Alarm, AlarmClient};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Led as usize;
+
+/// Holds the array of LEDs and implements a `Driver` interface to
+/// control them.
+pub struct LedMatrixDriver<'a, L: gpio::Pin, A: Alarm<'a>> {
+    cols: &'a [&'a L],
+    rows: &'a [&'a L],
+    buffer: TakeCell<'a, [u8]>,
+    alarm: &'a A,
+    current_row: Cell<usize>,
+    timing: u8,
+    row_activation: ActivationMode,
+    col_activation: ActivationMode,
+}
+
+impl<'a, L: gpio::Pin, A: Alarm<'a>> LedMatrixDriver<'a, L, A> {
+    pub fn new(
+        cols: &'a [&'a L],
+        rows: &'a [&'a L],
+        buffer: &'a mut [u8],
+        alarm: &'a A,
+        col_activation: ActivationMode,
+        row_activation: ActivationMode,
+        refresh_rate: usize,
+    ) -> Self {
+        // Initialize all LEDs and turn them off
+        if (buffer.len() * 8) < cols.len() * rows.len() {
+            panic!("Matrix LED Driver: provided buffer is too small");
+        }
+
+        Self {
+            cols,
+            rows,
+            buffer: TakeCell::new(buffer),
+            alarm,
+            col_activation: col_activation,
+            row_activation: row_activation,
+            current_row: Cell::new(0),
+            timing: (1000 / (refresh_rate * rows.len())) as u8,
+        }
+    }
+
+    pub fn init(&self) {
```

Comment:
If I understand correctly, it seems that this function toggles all the LEDs in the matrix in some pattern at some refresh rate? Should this be named something like `patterned_blink()`? Is anything in this function actually necessary to initialize the LED matrix?

Reply:
No, this function inits the rows and columns to 0 and/or 1 (depending on the settings). It basically turns off the matrix at the beginning.

Reply:
Ah, why does that require a timer?

Reply:
I thought you were referring to only to the init function. The driver does require a timer as the matrix can "display" one line at a time. The call to `next_row` just starts the marix functioning.

Reply:
Ah, I understand, thanks for explaining